sympy/series/formal.py

```diff
@@ -1464,6 +1455,260 @@ def __rmul__(self, other):
+        Term by term logic is implemented here.
+
+        Examples
+        ========
+
+        >>> from sympy import fps, sin, exp, convolution
+        >>> from sympy.abc import x
+        >>> f1 = fps(sin(x))
+        >>> f2 = fps(exp(x))
+        >>> fprod = f1.product(f2, x)
+
+        >>> fprod._eval_terms(4)
+        x**3/3 + x**2 + x
+
+        See Also
+        ========
+
+        sympy.series.formal.FormalPowerSeries.product
+
+        """
+        coeff1, coeff2 = self.ak
+
+        aks = convolution(coeff1[:n], coeff2[:n])
+
+        terms = []
+        for i in range(0, n):
+            terms.append(aks[i] * self.xk.coeff(i))
+
+        return Add(*terms)
+
+
+class FormalPowerSeriesCompose(FiniteFormalPowerSeries):
+    """Represents the composed formal power series of two functions.
+
+    No computation is performed. Terms are calculated using a term by term logic,
+    instead of a point by point logic.
+
+    There are two differences between a `FormalPowerSeries` object and a
+    `FormalPowerSeriesCompose` object. The first argument contains the outer
+    function and the inner function involved in the omposition. Also, the
+    coefficient sequence contains the generic sequence which is to be multiplied
+    by a custom `bell_seq` finite sequence. The finite terms will then be added up to
+    get the final terms.
+
+    See Also
+    ========
+
+    sympy.series.formal.FormalPowerSeries
+    sympy.series.formal.FiniteFormalPowerSeries
+
+    """
+    def __init__(self, *args):
+        k = self.xk.variables[0]
+        self.fact_seq = sequence(factorial(k), (k, 1, oo))
+        self.ffps, self.gfps = None, None
+
+    @property
+    def function(self):
+        """Function for the composed formal power series."""
+        f, g, x = self.f, self.g, self.x
+        return f.subs(x, g)
+
+    def _eval_terms(self, n):
+        """
+        Returns the first `n` terms of the composed formal power series.
+        Term by term logic is implemented here.
+
+        The coefficient sequence of the `FormalPowerSeriesCompose` object is the generic sequence.
+        It is multiplied by `bell_seq` to get a sequence, whose terms are added up to get
+        the final terms for the polynomial.
+
+        Examples
+        ========
+
+        >>> from sympy import fps, sin, exp, bell
+        >>> from sympy.abc import x
+        >>> f1 = fps(exp(x))
+        >>> f2 = fps(sin(x))
+        >>> fcomp = f1.compose(f2, x)
+
+        >>> fcomp._eval_terms(6)
+        -x**5/15 - x**4/8 + x**2/2 + x + 1
+
+        >>> fcomp._eval_terms(8)
+        x**7/90 - x**6/240 - x**5/15 - x**4/8 + x**2/2 + x + 1
+
+        See Also
+        ========
+
+        sympy.series.formal.FormalPowerSeries.compose
+        sympy.series.formal.FormalPowerSeries.coeff_bell
+
+        """
+        f, g = self.f, self.g
+
+        if self.ffps is None:
+            self.ffps = fps(f)
+        if self.gfps is None:
+            self.gfps = fps(g)
```

Comment:
This is saves computation time. But only problem is there is no way to maintain same arguments that was used to create the original ``fps(f)``.

As a suggestion, you can change the api of these objects to accept ``FormalPowerSeries`` argument except just the function names. That way using of cached ``FormalPowerSeries`` object will also me ensured.

eg.

``FormalPowerSeriesCompose((self, other), self.x, self.x0, self.dir, ((self.bell_coeff_seq), self.xk, self.ind))``

when returning for compose.

``self`` and ``other`` being FormalPowerSeries object have all the needed information.

Reply:
On second thought if you go by my way, you can get rid of so many arguments and just pass in

``FormalPowerSeriesCompose(self, other)``, since ``self`` and ``other`` already have all the information available.  This will make it much simpler.

Reply:
API has been changed to the one you suggested.